### PR TITLE
QML: Add support for Waveform Overviews

### PR DESCRIPTION
**Pull request by Holzhaus**
Implements a custom QQuickPaintItem that draws the waveform overviews as a QML component.

No support for markers, clicking, etc. has been added. IMHO this should be implemented using COs on the QML level for maximum flexibility.

Depends on #3911.

### CODE REVIEW

**Comment by Swiftb0y** [comment]
Could this also theoretically be used to replace the deprecated QGLWidget waveforms by embedding a QML engine in the skin which is used for just rendering the waveforms?

**Comment by Holzhaus** [comment]
> Could this also theoretically be used to replace the deprecated QGLWidget waveforms by embedding a QML engine in the skin which is used for just rendering the waveforms?

I don't know, I think the scrolling waveforms rely on some elaborate synchronization. But in theory, we could draw them using the same technique and scroll them in qml? No idea if this works latency-wise.

This is just about the waveform *overviews* (for quick seeking in the deck).

File: src/skin/qml/qmlplayermanagerproxy.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+class QmlPlayerManagerProxy : public QObject {
+    Q_OBJECT
+  public:
+    explicit QmlPlayerManagerProxy(PlayerManager* pPlayerManager, QObject* parent = nullptr);
+
+    Q_INVOKABLE QObject* getPlayer(const QString& deck);
+
+  private:
+    const PlayerManager* m_pPlayerManager;
```
Comment (uklotzde) [inline src/skin/qml/qmlplayermanagerproxy.h:19]: ~~I suggest to use `const QPointer<PlayerManager>` that gets reset implicitly.~~

-> `const std::shared_ptr<PlayerManager>`

File: src/skin/qml/qmlplayerproxy.h [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    void groupingChanged();
+    void yearChanged();
+    void trackNumberChanged();
+    void trackTotalChanged();
+    void commentChanged();
+    void keyTextChanged();
+    void colorChanged();
+
+  private:
+    BaseTrackPlayer* m_pTrackPlayer;
```
Comment (uklotzde) [inline src/skin/qml/qmlplayerproxy.h:94]: Also use `const QPointer<BaseTrackPlayer>`?

File: src/skin/qml/qmlskin.cpp [hunk 3]
```diff
@@ -104,12 +119,34 @@ QWidget* QmlSkin::loadSkin(QWidget* pParent,
...
+            0,
+            1,
+            "PlayerManager",
+            lambda_to_singleton_type_factory_ptr(
+                    [pCoreServices](QQmlEngine* pEngine,
+                            QJSEngine* pScriptEngine) -> QObject* {
+                        Q_UNUSED(pScriptEngine);
+
+                        QmlPlayerManagerProxy* pPlayerManagerProxy =
+                                new QmlPlayerManagerProxy(
```
Comment (uklotzde) [inline src/skin/qml/qmlskin.cpp:139]: Passing the plain pointer that is actually managed by shared_ptr<> and not by Qt is dangerous! Then we should instead consistently use std::shared_ptr<PlayerManager>.

File: src/skin/qml/qmlwaveformoverview.h [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            Channels channels,
+            ConstWaveformPointer pWaveform,
+            int completion) const;
+    void drawRgb(QPainter* pPainter,
+            Channels channels,
+            ConstWaveformPointer pWaveform,
+            int completion) const;
+    QColor getRgbPenColor(ConstWaveformPointer pWaveform, int completion) const;
+
+    QmlPlayerProxy* m_pPlayer;
```
Comment (uklotzde) [inline src/skin/qml/qmlwaveformoverview.h:75]: `const QPointer<QmlPlayerProxy>`?

**Review by uklotzde — CHANGES_REQUESTED** [review]

File: src/skin/qml/qmlplayermanagerproxy.h (hunk 1 above)
Comment (uklotzde) [reply src/skin/qml/qmlplayermanagerproxy.h:19]: Do we need a typedef then?

File: src/skin/qml/qmlskin.cpp (hunk 3 above)
Comment (Holzhaus) [reply src/skin/qml/qmlskin.cpp:139]: No, instead we should tell the Qml Engine that it's managed by the JS engine.

File: src/skin/qml/qmlwaveformoverview.cpp [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+    return m_pPlayer;
+}
+
+void QmlWaveformOverview::setPlayer(QmlPlayerProxy* pPlayer) {
+    if (m_pPlayer == pPlayer) {
+        return;
+    }
+
+    if (m_pPlayer != nullptr) {
+        disconnect(m_pPlayer->internalTrackPlayer(), nullptr, this, nullptr);
```
Comment (uklotzde) [inline src/skin/qml/qmlwaveformoverview.cpp:34]: Why not just `m_pPlayer->internalTrackPlayer()->disconnect(this)`?

https://doc.qt.io/qt-5/qobject.html#disconnect-3

File: src/skin/qml/qmlwaveformoverview.h (hunk 4 above)
Comment (uklotzde) [reply src/skin/qml/qmlwaveformoverview.h:75]: Ping

Comment (Holzhaus) [reply src/skin/qml/qmlwaveformoverview.h:75]: This can't be const. I'll switch to QPointer.

File: src/skin/qml/qmlwaveformoverview.cpp (hunk 5 above)
Comment (Holzhaus) [reply src/skin/qml/qmlwaveformoverview.cpp:34]: good idea, didn't know this method.

**Review by uklotzde — APPROVED** [review]

**Comment by uklotzde** [comment]
LGTM

**Comment by Be-ing** [comment]
I think this should be rewritten... https://doc.qt.io/qt-6/qtquick-visualcanvas-scenegraph.html#custom-items-using-qpainter

> Custom Items using QPainter
The QQuickItem provides a subclass, QQuickPaintedItem, which allows the users to render content using QPainter.
Warning: Using QQuickPaintedItem uses an indirect 2D surface to render its content, either using software rasterization or using an OpenGL framebuffer object (FBO), so the rendering is a two-step operation. First rasterize the surface, then draw the surface. Using scene graph API directly is always significantly faster.

**Comment by Swiftb0y** [comment]
Once we have a stable solution for the scrolling waveforms, we could use the same for the overviews...

**Comment by Holzhaus** [comment]
Sure, we could use a shader.

**Comment by Holzhaus** [comment]
On second thought, are shaders redrawn only on changes? If not, we should probably move the overviews to the image provider. After analysis, the don't change anymore (in contrast to scroll waveforms that react to eq changes)

**Comment by Swiftb0y** [comment]
Mhmm, true. Though it does change when changing the overview type...

**Pull request merged**